src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs

```diff
@@ -76,9 +93,39 @@ public BaseConfiguration LastKnownGoodConfiguration
             {
                 _lastKnownGoodConfiguration = value ?? throw LogHelper.LogArgumentNullException(nameof(value));
                 _lastKnownGoodConfigFirstUse = DateTime.UtcNow;
+
+                if (_lkgConfigurationCache == null)
+                    _lkgConfigurationCache = new Dictionary<BaseConfiguration, DateTime>(BaseConfigurationComparer);
+                
+                _lkgConfigurationCache[_lastKnownGoodConfiguration] = DateTime.UtcNow + LastKnownGoodLifetime;
+
+                //remove expired configuration to avoid memory leak 
+                foreach (KeyValuePair<BaseConfiguration, DateTime> lkgConfiguration in _lkgConfigurationCache)
+                {
+                    if (lkgConfiguration.Value < DateTime.UtcNow)
+                        _lkgConfigurationCache.Remove(lkgConfiguration.Key);
+                }
             }
         }
 
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <returns></returns>
+        public IDictionary<BaseConfiguration, DateTime> GetValidLkgConfiguraitonFromCache()
+        {
+            if (_lkgConfigurationCache == null)
+                return null;
+
+            foreach (KeyValuePair<BaseConfiguration, DateTime> lkgConfiguration in _lkgConfigurationCache)
```

Comment:
>foreach [](http://example.com/codeflow?start=12&length=7)

same comment regarding removing while iterating, consider a private helper since you're doing this in two places.